### PR TITLE
Frontend changes to enable multi content annotations

### PR DESCRIPTION
**Pull request by luniki**
This pull request will contain all changes required to enable multi content annotations.

What does this necessitate:

* [X] Change the data model
* [X] Add a way to add an empty multi content annotation
* [x] Change the way annotations are shown in the list of annotations
* [x] Enable adding new content to a multi content annotation
* [x] Enable dropping content from a multi content annotation
* [x] Enable editing content of a multi content annotation
* [x] Change the way annotations are shown in the timeline
* [x] Refactor existing code to use the new data structuring

Refs #370

### CODE REVIEW

**Comment by luniki** [comment]
I will have to clean-up the code. But more or less this should do it.

Prepare for reviewing, please! :rofl:

**Comment by luniki** [comment]
OK lets do this

File: frontend/templates/list-annotation-category.tmpl [hunk 1]
```diff
@@ -0,0 +1,71 @@
+{{!-- Skalenwerte zuerst --}}
+{{#each content}}
+    {{#ifeq type "scaling"}}
+        <span class="scaling">
+            {{#with (withScaleValue value.scaling) as |scaleValue|}}
+                <span class="read-only" title="{{scaleValue.name}}" >{{scaleValue.name}}</span>
```
Comment (JulianKniephoff) [inline frontend/templates/list-annotation-category.tmpl:6]:
```suggestion
                <span class="read-only" title="{{scaleValue.name}}">{{scaleValue.name}}</span>
```

File: frontend/templates/list-annotation-category.tmpl [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            {{/with}}
+        </span>
+    {{/ifeq}}
+{{/each}}
+
+{{#each content}}
+
+    {{!-- Annotationen mit Kennzeichen und Skalenwert --}}
+    {{#ifeq type "scaling"}}
```
Comment (JulianKniephoff) [inline frontend/templates/list-annotation-category.tmpl:15]: This whole `ifeq` is indented weirdly. Also as I noted elsewhere, I would indent the contents of "block helpers" as well, so

```
{{ifeq ...}}
    <span>...</span>
{{/ifeq}}
```

as opposed to 
```
{{ifeq ...}}
<span>...</span>
{{/ifeq}}
```

File: frontend/templates/list-annotation-category.tmpl [hunk 3]
```diff
@@ -0,0 +1,71 @@
+{{!-- Skalenwerte zuerst --}}
```
Comment (JulianKniephoff) [inline frontend/templates/list-annotation-category.tmpl:1]: The name of this partial seems weird to me. Shouldn't it rather be `list-annotation-content`? Of course this also applies to the variables bound to it.

File: frontend/templates/list-annotation-expanded.tmpl [hunk 4]
```diff
@@ -96,13 +66,28 @@
...
+        <div>
+            <span>{{t "annotation.add content.list prompt"}}</span>
+            <button class="btn add-free-text-content" type="button">
+                {{t "annotation.add content.free text"}}
+            </button>
+
+            {{#each categories}}
+            <button class="btn add-labelled-content" type="button" data-category="{{id}}" {{#if settings.color}}style="background-color:{{settings.color}};background-image: none;"{{/if}}>
+                {{t "annotation.add content.category"}} {{name}}
+            </button>
```
Comment (JulianKniephoff) [inline frontend/templates/list-annotation-expanded.tmpl:86]:
```suggestion
                <button
                    class="btn add-labelled-content"
                    type="button"
                    data-category="{{id}}"
                    {{#if settings.color}}
                        style="
                            background-color: {{settings.color}};
                            background-image: none;
                        "
                    {{/if}}
                >
                    {{t "annotation.add content.category"}} {{name}}
                </button>
```

File: frontend/templates/content-item-header.tmpl [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+        {{> @partial-block }}
+    </span>
+    <span class="actions">
+        {{#if annotation.isMine}}
+        <i class="icon-pencil content-item-edit"
+           title="{{t "annotation.edit.edit"}}"
+        ></i>
+        <i class="icon-trash content-item-trash"
+           title="{{t "annotation.edit.delete"}}"
+        ></i>
```
Comment (JulianKniephoff) [inline frontend/templates/content-item-header.tmpl:16]:
```suggestion
            <i
                class="icon-pencil content-item-edit"
                title="{{t "annotation.edit.edit"}}"
            ></i>
            <i
                class="icon-trash content-item-trash"
                title="{{t "annotation.edit.delete"}}"
            ></i>
```

File: frontend/templates/content-type-label.tmpl [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+            <b>{{title}}</b>
+        {{^}}
+            <b> {{label.category.name}} </b>
+            <i>({{t "annotation.types.label"}})</i>
+        {{/if}}
+    {{/content-item-header}}
+    <div class="well">
+        <span class="text">{{label.category.name}} - {{label.value}} ({{label.abbreviation}})</span>
+    </div>
+</div>
```
Comment (JulianKniephoff) [inline frontend/templates/content-type-label.tmpl:14]:
```suggestion
    <div class="content-item" data-position="{{@index}}"{{#if category.settings.color}} style="background-color: {{category.settings.color}}"{{/if}}>
        {{#> content-item-header annotation=../annotation}}
            {{#if title}}
                <b>{{title}}</b>
            {{^}}
                <b> {{label.category.name}} </b>
                <i>({{t "annotation.types.label"}})</i>
            {{/if}}
        {{/content-item-header}}
        <div class="well">
            <span class="text">{{label.category.name}} - {{label.value}} ({{label.abbreviation}})</span>
        </div>
    </div>
```

File: frontend/templates/content-type-scaling.tmpl [hunk 7]
```diff
@@ -0,0 +1,34 @@
+{{#with (withLabel value.label) as |label|}}
```
Comment (JulianKniephoff) [inline frontend/templates/content-type-scaling.tmpl:1]: This, too, I would reindent like indicated in many of the other templates.

File: frontend/js/views/list-annotation.js [hunk 8]
```diff
@@ -581,6 +536,99 @@ define(["jquery",
...
                 Backbone.View.prototype.remove.apply(this, arguments);
+            },
+
+            /**
+             * Add a new annotation with a content item containing free text.
+             * @alias module:views-list-annotation.ListAnnotation#addFreetextContent
+             * @param {Event} event Event object
+             */
+            addFreeTextContent(event) {
+                event && event.stopPropagation();
```
Comment (JulianKniephoff) [inline frontend/js/views/list-annotation.js:547]:
```suggestion
                event.stopPropagation();
```

Is this ever called without event?

File: frontend/js/views/list-annotation.js [hunk 9]
```diff
@@ -581,6 +536,99 @@ define(["jquery",
...
+                );
+            },
+
+            /**
+             * Add a new annotation with a content item containing a label w/ or w/o a scale value.
+             * @alias module:views-list-annotation.ListAnnotation#addLabelledContent
+             * @param {Event} event Event object
+             */
+            addLabelledContent(event) {
+                event && event.stopPropagation();
```
Comment (JulianKniephoff) [inline frontend/js/views/list-annotation.js:562]:
```suggestion
                event.stopPropagation();
```

File: frontend/js/annotation-tool.js [hunk 10]
```diff
@@ -24,14 +24,15 @@ define(["jquery",
         "i18next",
         "collections/videos",
         "views/main",
+        "views/modal-container",
         "alerts",
         "templates/delete-modal",
         "templates/delete-warning-content",
         "player-adapter",
         "colors",
         "handlebarsHelpers"],
```
[the remark this answers is not in the document]
Comment (luniki) [reply frontend/js/annotation-tool.js:33]: Are you sure? AFAIK I only "reformatted" them in new files (and it wasn't me but prettier).

File: frontend/js/annotation-tool.js [hunk 11]
```diff
@@ -24,14 +24,15 @@ define(["jquery",
...
 
-    function ($, _, Backbone, i18next, Videos, MainView, alerts, DeleteModalTmpl, DeleteContentTmpl, PlayerAdapter, ColorsManager) {
+       function ($, _, Backbone, i18next, Videos, MainView, ModalContainerView, alerts, DeleteModalTmpl, DeleteContentTmpl, PlayerAdapter, ColorsManager) {
```
[the remark this answers is not in the document]
Comment (luniki) [reply frontend/js/annotation-tool.js:35]: So what's the rule here: The "define" function shall not indent the first argument but the second argument must be indented by 4 spaces?

File: frontend/js/annotation-tool.js [hunk 12]
```diff
@@ -794,6 +795,20 @@ define(["jquery",
                         }
                     }, this)
                 );
+            },
+
+            addModal: function (header, contentView, buttonText) {
```
[the remark this answers is not in the document]
Comment (luniki) [reply frontend/js/annotation-tool.js:800]: The idea is that not every view has to import an implementation detail but just use a method of the globally available `annotationTool`.

File: frontend/js/collections/annotation_content.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
+/**
```
[the remark this answers is not in the document]
Comment (luniki) [reply frontend/js/collections/annotation_content.js:1]: Ok

File: frontend/js/collections/annotation_content.js [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+     * @constructor
+     * @see {@link http://www.backbonejs.org/#Collection}
+     * @augments module:Backbone.Collection
+     * @memberOf module:collections-annotation-content
+     * @alias module:collections-annotation-content.AnnotationContent
+     */
+    var AnnotationContent = Backbone.Collection.extend({
+        /**
+         * Model of the instances contained in this collection
+         * @alias module:collections-annotation-content.AnnotationContent#initialize
```
[the remark this answers is not in the document]
Comment (luniki) [reply frontend/js/collections/annotation_content.js:38]: Almost every single file in `js/collections` but `tracks.js` uses `initialize`.

File: frontend/js/handlebarsHelpers.js [hunk 15]
```diff
@@ -73,5 +73,34 @@ define(["handlebars", "underscore", "i18next", "util"], function (Handlebars, _,
...
+    /**
+     * Useful for debugging templates.
+     * @alias module:Handlebars#json
+     */
+    Handlebars.registerHelper("json", function (value) {
+        return new Handlebars.SafeString(
+            JSON.stringify(value)
+        );
+    });
+
```
[the remark this answers is not in the document]
Comment (luniki) [reply frontend/js/handlebarsHelpers.js:85]: In which way cleaner?

File: frontend/js/models/annotation.js [hunk 16]
```diff
@@ -71,6 +76,12 @@ define(["underscore",
                     this.attributes.comments = attr.comments;
                     delete attr.comments;
                 }
+
+                if (_.isObject(attr.content)) {
+                    this.attributes.content = attr.content;
+                } else {
+                    this.attributes.content = new AnnotationContent(attr.content || []);
+                }
```
[the remark this answers is not in the document]
Comment (luniki) [reply frontend/js/models/annotation.js:84]: IMHO all this cruft should be removed at once and currently we do not have the necessary time to do it. This is consistent and it works.

File: frontend/templates/modal-edit-free-text.tmpl [hunk 17]
```diff
@@ -0,0 +1,14 @@
+{{#if contentItem.title}}
+<header><h2>{{contentItem.title}}</h2></header>
+{{/if}}
```
[the remark this answers is not in the document]
Comment (luniki) [reply frontend/templates/modal-edit-free-text.tmpl:3]: Content items do have titles.

**Comment by JulianKniephoff** [comment]
Closing in favor of #449.

**Pull request closed without merging**